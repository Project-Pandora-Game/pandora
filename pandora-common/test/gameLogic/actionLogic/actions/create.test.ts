import { describe, expect, it } from '@jest/globals';
import { Assert, ParseNotNullable } from '../../../../src/index.ts';
import { TestAssetsLoadAssetManager } from '../../../assets/_testData/testAssetsDefinition.ts';
import { TestCreateCharacterState, TestCreateGlobalState } from '../../../assets/assetsTestingHelpers.ts';
import { TestCreateGameLogicCharacter } from '../../character/characterTestingHelpers.ts';
import { TestActionExpectValidResult, TestCreateActionContext, TestDoImmediateAction, TestStateExtractAssets } from '../actionTestingHelpers.ts';

describe('ActionCreate', () => {
	it('Creates a non-unique bodypart', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room),
		]);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'create',
				target: {
					type: 'character',
					characterId: character.id,
				},
				container: [],
				itemTemplate: {
					asset: 'a/body/front_hair1',
				},
			},
			TestCreateActionContext(character, { spaceFeatures: ['allowBodyChanges'] }),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Creates a non-unique bodypart and reorders it in place', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');

		// Add front hair, as back hair should be inserted below front hair
		const frontHair = assetManager.createItem(
			`i/${character.id}:front_hair1`,
			ParseNotNullable(assetManager.getAssetById('a/body/front_hair1')),
			character,
		);
		Assert(frontHair.isType('bodypart'));
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => {
				const baseCharacterState = TestCreateCharacterState(assetManager, character, room);
				return baseCharacterState.produceWithItems([
					...baseCharacterState.items,
					frontHair,
				]);
			},
		]);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'create',
				target: {
					type: 'character',
					characterId: character.id,
				},
				container: [],
				itemTemplate: {
					asset: 'a/body/back_hair_normal',
				},
			},
			TestCreateActionContext(character, { spaceFeatures: ['allowBodyChanges'] }),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Creates a unique bodypart and replaces the old one', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room),
		]);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'create',
				target: {
					type: 'character',
					characterId: character.id,
				},
				container: [],
				itemTemplate: {
					asset: 'a/body/eyes2',
				},
			},
			TestCreateActionContext(character, { spaceFeatures: ['allowBodyChanges'] }),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Creates a wearable asset', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room),
		]);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'create',
				target: {
					type: 'character',
					characterId: character.id,
				},
				container: [],
				itemTemplate: {
					asset: 'a/panties/style1',
				},
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Creates a wearable asset and another one ordered before it', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room),
		]);

		// Do the actions
		const intermediateResult = TestDoImmediateAction(
			{
				type: 'create',
				target: {
					type: 'character',
					characterId: character.id,
				},
				container: [],
				itemTemplate: {
					asset: 'a/panties/style1',
				},
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the intermediate result
		TestActionExpectValidResult(intermediateResult);

		const items = intermediateResult.resultState.getItems({
			type: 'character',
			characterId: character.id,
		});

		const result = TestDoImmediateAction(
			{
				type: 'create',
				target: {
					type: 'character',
					characterId: character.id,
				},
				container: [],
				itemTemplate: {
					asset: 'a/headwear/top_hat',
				},
				insertBefore: items?.[items.length - 1].id,
			},
			TestCreateActionContext(character),
			intermediateResult.resultState,
		);

		// Check the final result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});
});
