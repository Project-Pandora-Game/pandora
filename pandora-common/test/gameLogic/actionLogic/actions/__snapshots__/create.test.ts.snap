// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ActionCreate Creates a non-unique bodypart 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/body/front_hair1",
  },
]
`;

exports[`ActionCreate Creates a non-unique bodypart and reorders it in place 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/body/back_hair_normal",
  },
  {
    "asset": "a/body/front_hair1",
  },
]
`;

exports[`ActionCreate Creates a unique bodypart and replaces the old one 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes2",
  },
  {
    "asset": "a/body/lips",
  },
]
`;

exports[`ActionCreate Creates a wearable asset 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/panties/style1",
  },
]
`;

exports[`ActionCreate Creates a wearable asset and another one ordered before it 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/headwear/top_hat",
  },
  {
    "asset": "a/panties/style1",
  },
]
`;
